File: strStr.py
"""
Implement strStr().

Return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
"""
def strStr(haystack, needle):
    """
    :type haystack: str
    :type needle: str
    :rtype: int
    """
    if(len(haystack)> 0 and len(needle)>0):
        if(len(haystack) >= len(needle)):
            ans = -1
            i=0
            while(i<=len(haystack)-len(needle)):
                if needle[0] == haystack[i]:
                    j = i
                    ans = j
                    for k in range(len(needle)):
                        if needle[k] == haystack[j]:
                            j +=1
                        else:
                            break
                        if k == len(needle) - 1:
                            return ans
                i+=1
            return -1
        else:
            return -1
    elif(len(haystack)==0 and len(needle)>0):
        return -1
    else:
        return 0

File: strStr.cpp
class Solution {
public:
    
    int strStr(string haystack, string needle) {
        int n = haystack.length();
        int m = needle.length();
        if(m == 0){
            return m;
        }
        // preprocessing
        vector<int> lps(m, 0);
        for(int i = 1;i<m;i++){
            int pos = lps[i-1];
            while(pos>0 && needle[pos]!=needle[i]){
                pos = lps[pos-1];
            }
            if(needle[pos] == needle[i]){
                lps[i] = pos + 1;
            }
        }
        int idx_haystack = 0;
        int idx_needle = 0;
        for(idx_haystack = 0;idx_haystack<n;idx_haystack++){
            while(idx_needle>0 && needle[idx_needle]!=haystack[idx_haystack]){
                idx_needle = lps[idx_needle-1];
            }
            if(needle[idx_needle] == haystack[idx_haystack]){
                idx_needle++;
            }
            if(idx_needle == m){
                return idx_haystack - m + 1;
            }
        }
        return -1;
        
    }
};